selectors/onboarding.ts

```diff
@@ -34,3 +33,92 @@ export function getFirstChannelName(state: GlobalState) {
+    roles: string[];
+};
+
+export const Steps: StepType[] = [
+    {
+        id: RecommendedNextSteps.COMPLETE_PROFILE,
+        roles: [],
+    },
+    {
+        id: RecommendedNextSteps.TEAM_SETUP,
+        roles: ['first_admin'],
+    },
+    {
+        id: RecommendedNextSteps.NOTIFICATION_SETUP,
+        roles: ['system_user'],
+    },
+    {
+        id: RecommendedNextSteps.PREFERENCES_SETUP,
+        roles: ['system_user'],
+    },
+    {
+        id: RecommendedNextSteps.INVITE_MEMBERS,
+        roles: ['system_admin', 'system_user'],
+    },
+    {
+        id: RecommendedNextSteps.DOWNLOAD_APPS,
+        roles: [],
+    },
+];
+
+// Filter the steps shown by checking if our user has any of the required roles for that step
+export function isStepForUser(step: StepType, roles: string): boolean {
+    const userRoles = roles?.split(' ');
+    return (
+        userRoles?.some((role) => step.roles.includes(role)) ||
+          step.roles.length === 0
+    );
+}
+
+const getSteps = createSelector(
+    'getSteps',
+    (state: GlobalState) => getCurrentUser(state),
+    (state: GlobalState) => isFirstAdmin(state),
+    (currentUser, firstAdmin) => {
+        const roles = firstAdmin ? `first_admin ${currentUser?.roles}` : currentUser?.roles;
+        return Steps.filter((step) => isStepForUser(step, roles));
+    },
+);
+
+// show Onboarding task list if they haven't skipped/finished the legacy steps or there are legacy steps unfinished
+export const showOnboardingTaskListToExistingUsers = createSelector(
+    'showOnboardingTaskListToExistingUsers',
+    (state: GlobalState) => getCategory(state, Preferences.RECOMMENDED_NEXT_STEPS),
+    (state: GlobalState) => legacyNextStepsNotFinished(state),
+    (stepPreferences, legacyNextStepsNotFinished) => {
+        if (stepPreferences.some((pref) => (pref.name === RecommendedNextSteps.SKIP && pref.value === 'true'))) {
+            return false;
+        }
+
+        if (stepPreferences.some((pref) => (pref.name === RecommendedNextSteps.HIDE && pref.value === 'true'))) {
+            return false;
+        }
+
+        return legacyNextStepsNotFinished;
+    },
+);
+
+// Loop through all Steps. For each step, check that
+export const legacyNextStepsNotFinished = createSelector(
+    'legacyNextStepsNotFinished',
+    (state: GlobalState) => getCategory(state, Preferences.RECOMMENDED_NEXT_STEPS),
+    (state: GlobalState) => getCurrentUser(state),
+    (state: GlobalState) => isFirstAdmin(state),
+    (state: GlobalState) => getSteps(state),
+    (stepPreferences, currentUser, firstAdmin, mySteps) => {
+        const roles = firstAdmin ? `first_admin ${currentUser?.roles}` : currentUser?.roles;
+        const checkPref = (step: StepType) => stepPreferences.some((pref) => (pref.name === step.id && pref.value === 'true') || !isStepForUser(step, roles));
+        return !mySteps.every(checkPref);
+    },
+);
```

Comment:
Do we need to check for each step still ? It seems that these don't correlate to the new onboarding list

Reply:
yes, they don't correlate, but it makes sense that if a user hasn't finish the previous steps to see the new task list. Could potentially help end users to learn new things and admins to get the motivation to start a trial.